### ISSUE TITLE
Cannot drag-&-drop sources, dragging scenes deletes them when using browser panels

### ISSUE THREAD

**Issue by xurei** [obsproject/obs-studio#4488]
### Operating System Info

Other

### Other OS

Manjaro Linux

### OBS Studio Version

27.0.0-rc1

### OBS Studio Log URL

https://obsproject.com/logs/sw-7ofvWAMM1fHMa

### Expected Behavior

On OBS 26, it was possible to move sources in the sources list with drag-and-drop.

It was also the only way (that I know of) to put a source in an existing group.

### Current Behavior

On OBS 27, it does not work anymore.

### Steps to Reproduce

[EDIT 2021-05-14] I made a video highlighting the problem : https://www.youtube.com/watch?v=QGDVaOUOr1s

Moving sources
1. Create a blank scene with two sources (whichever should work)
2. Try to move the upper source down using drag-and-drop
3. It should not work

Putting sources in group
1. Create a blank scene with a source and a group
2. Try to put the source in the group using drag-and-drop
3. It should not work

### Anything else we should know?

I've taken a look in the logs and didn't see any kind of error...

**Comment by WizardCM** [comment]
I can confirm this on Ubuntu 20.04. Worse, if you try and drag scenes after reproducing this issue, it straight up deletes the scene.

**Comment by awarus** [comment]
Ubuntu 18.04 5.4.0-70-generic #78~18.04.1-Ubuntu SMP is OK (with both 27.0.0-rc1 and 27.0.0-rc2) with different scenarios. May affect some libqt on this(my version is 5)? Like manipulating in UI with QDrag objects and modes?

**Comment by RytoEX** [comment]
Ubuntu 18.04 has Qt 5.9.5.  Ubuntu 20.04 has Qt 5.12.8.  I believe Manjaro should have Qt 5.15.2.

**Comment by xurei** [comment]
@RytoEX Confirmed : Qt 5.15.2

```
$ qtdiag
Qt 5.15.2 (x86_64-little_endian-lp64 shared (dynamic) release build; by GCC 10.2.0) on "xcb" 
OS: Manjaro Linux [linux version 5.10.23-1-MANJARO]
...
```

**Comment by RytoEX** [comment]
For what it's worth, so far I cannot reproduce this on Windows 10 with OBS Studio 27.0.0-rc1 or 27.0.0-rc2.  I also cannot reproduce this in a Ubuntu 20.04 VM with OBS Studio 27.0.0-rc2.

**Comment by WizardCM** [comment]
After performing a bisect to try and track down when this issue was introduced, `git bisect` told me the following:

```
90ab876cc727a1d9ef89a32f065fa25efc17f98 is the first bad commit
commit 190ab876cc727a1d9ef89a32f065fa25efc17f98
Merge: e8a1c082d e67fdbca7
Author: Jim <obs.jim@gmail.com>
Date:   Fri Feb 12 01:30:05 2021 -0800

    Merge pull request #2484 from GeorgesStavracas/feaneron/egl-wayland
    
    EGL/Wayland (Wayland, pt 3)
```

However, now that I've finished the bisect & after returning HEAD to `master`, I can no longer reproduce this issue.

I can only assume that either
1. a stale built file stuck around from somewhere and introduced the issue
2. a stale cmake flag stuck around from somewhere and introduced the issue

Would you be willing to do the following?

1. Clone a second copy of the `obs-studio` repo and try building it clean from current `master`

And if that doesn't have the same issue, would you be able to dump your cmake config/flags from the "broken" variant?

I'd do it myself but wasn't expecting the issue to magically disappear and I now have no way to reproduce this issue.

**Comment by RytoEX** [comment]
@xurei How did you install this version of OBS Studio?  What variant and version of Manjaro are you running?

**Comment by kkartaltepe** [comment]
I also cannot replicate on Arch Linux, on-top of 00fdec2493479de24f6669ce26f5ba3f5383ce78

**Comment by RytoEX** [comment]
I'm leaning towards some scene collection file/data corruption, possibly from undo/redo, and that corruption snowballing into other issues like this.  I think it's possible that this issue has since been resolved (or perhaps simply masked) by code changes made since then.

As far as I can tell, "OBS 27.0.0.rc1.r26.gc37069277" would be commit https://github.com/obsproject/obs-studio/commit/c37069277335d5cb445792e3a062a0512dbd0307.  Here is the comparison between that commit and RC2:  https://github.com/obsproject/obs-studio/compare/c37069277...27.0.0-rc2

In that diff, you'll see the following:
* https://github.com/obsproject/obs-studio/commit/bab21888f389cf16c2966d9c7a099682dd6f4974
* https://github.com/obsproject/obs-studio/commit/2932cd0fd3b83e6b07192a784bf3faebdebaa697

There are even more changes since then.  I'd recommend:
1. Testing on commit https://github.com/obsproject/obs-studio/commit/c37069277335d5cb445792e3a062a0512dbd0307 with a clean scene collection to see if it's able to be reproduced there to validate these reproduction steps.
2. Testing on the current master (or RC3 when it's out) with a clean scene collection to see if it can still be reproduced there.

**Comment by xurei** [comment]
@RytoEX It's installed it through AUR (https://aur.archlinux.org/packages/obs-studio-git/). I just made an update and the problem is still there.
Version installed : 27.0.0.rc3.r15.g54db450e4

**Comment by rogue-ronin** [comment]
Saw a link to this in discord chat, thought I'd check it out. I just installed _obs-studio-git_ in Manjaro Linux (27.0.0.rc3.r48.gca88ba8bb)

I can drag-n-drop, move into group, move out of group, reorder in group, reorder group within Sources list. No errors at all.

I tried this both before and after updating Manjaro to the latest weekly version.
Kernel: 5.12.1-2-MANJARO

So... :woman_shrugging: perhaps it's local?

**Comment by xurei** [comment]
@rogue-ronin I just updated OBS to 27.0.0.rc4.r3.g5fcffb066
The problem is still present.

What is the version of Qt that you get with `qtdiag` ?

**Comment by RytoEX** [comment]
> 
> 
> @rogue-ronin I just updated OBS to 27.0.0.rc4.r3.g5fcffb066
> The problem is still present.
> 
> What is the version of Qt that you get with `qtdiag` ?

Can you please clarify if this is reproducible with a clean scene collection?

**Comment by xurei** [comment]
@RytoEX Just tried. Yes it is reproducible.
I made a video highlighting the problem : https://www.youtube.com/watch?v=QGDVaOUOr1s

**Comment by RytoEX** [comment]
> 
> 
> @RytoEX Just tried. Yes it is reproducible.
> I made a video highlighting the problem : https://www.youtube.com/watch?v=QGDVaOUOr1s

Please reproduce in a clean installation of OBS Studio.  I see StreamFX installed.

**Comment by xurei** [comment]
@RytoEX I got it working !

Here is what I did :
1. Uninstalled every OBS plugin
2. Moved the ~/.config/obs-studio folder to a temporary folder
3. Started a fresh OBS
4. Reinstalled the plugins I removed
5. Put back in ~/.config/obs-studio/basic folder from the saved folder to get back my scenes.

What I understand is that some configuration was interfering somehow, but I couldn't find what exactly...

Thank you for pointing out that it could be the cause.

**Comment by rogue-ronin** [comment]
Good to hear. Glad I could catalyze. :)

**Comment by mihawk90** [comment]
I may or may not have found a way to reproduce this reliably (and yes I know it sounds insane).

This bug seems to happen anytime a browser dock is loaded from scratch (enabling and disabling in the View menu only seems to toggle the visibility, not actually close/unload it).

Can someone please try the following steps:

0. Delete/move `~/.config/obs-studio` for a clean new completely default config
1. Start OBS (skip wizard, shouldn't matter)
2. Create 2 scenes & move them (should work)
3. Create custom browser dock (for me: Name = Custom, URL = https://obsproject.com/browser-source)
4. Dock the Custom Browser Dock (for me it was to the right of the Program/Preview)
5. Move scenes (should work)
6. (optional) Hide Dock via View menu & move scenes (should work)
7. (optional) Show Dock via View menu & move scenes (should work)
8. Close Browser Dock *via the X on it* **not** through the View menu
9. Move scenes (should work)
10. Enable Browser Dock via View menu
11. Move scenes => Success(?) - **Moved scene disappears**
12. Close Browser Dock (again via X)
13. Move scenes (should work)

If in Step 12 you only hide the Dock via the View menu you will notice that in 13 the scene instead disappears.

It seems to be in some way related loading the custom docks.

**Comment by RytoEX** [comment]
Even with the detailed steps above, I am unable to reproduce this on Windows with OBS Studio 27.0.0-rc5.

**Comment by WizardCM** [comment]
I can confirm 100% that this seems to be related to browser docks per @mihawk90's steps. Closing all browser docks fixes the issue. It only happens on Linux, not macOS or Windows. Verified on Ubuntu 20.04 (Qt 5.12.8)

I have no idea where to go from here to be honest, as the existence of a browser dock should not affect drag/drop data - at least not directly.

**Comment by mihawk90** [comment]
Should add:
I was testing on Fedora 33 KDE with Qt 5.15.2

On a side note: We might want to update the Issue title since it is not just about Sources, and I think deleting scenes is more important anyway (especially when people just check the 27 RC Bug Tracker project for known/remaining issues).

**Comment by RytoEX** [comment]
I've done some more testing, and I can confirm that I can reproduce this on Ubuntu 21.04 with OBS 27.0.0-rc5 with the steps from @mihawk90 in [the comment above](https://github.com/obsproject/obs-studio/issues/4488#issuecomment-846266434).

The reproduction steps can be simplified just a bit:

1. Delete/move ~/.config/obs-studio for a clean new completely default config.
2. Start OBS.  Skip the auto config wizard.
3. Create 2 scenes.
4. Create custom browser dock. The name and URL don't matter, but here are sample values:
    (Name = Custom, URL = https://obsproject.com/browser-source)
5. Dock the Custom Browser Dock anywhere in the main window.
6. Close Browser Dock with the X button on it not through the View menu.  Dismiss the message box about closing docks.
    Up until this point, you can drag and drop scenes with no problem.
7. Enable Browser Dock via the View menu.
8. Drag and drop a scene to move it.
    **The scene disappears from UI and the scene collection JSON.**
9. Close Browser Dock with its X button.
    Dragging and dropping a scene works OK again at this point.

Observations:
1. The issue manifests differently for sources and scenes.
    Sources become unable to be reordered via drag and drop.  They visually look like they are being dragged, but their order does not change once dropped.
    Scenes can be dragged, but the dragged scene will disappear once dropped.
2. The issue does not manifest with the move up and move down buttons.  It only manifests if you drag and drop an item.
3. The issue will also manifest if you start up OBS with a docked browser dock, including service related browser docks/panels (e.g., Twitch Chat).
4. Once manifested, the issue goes away if you _close_ the browser dock by its X button or if you undock the Browser Dock and restart OBS.  Hiding it via the View -> Docks menu does not make the issue go away.

The key seems to be a docked browser dock/panel being created in a docked state.  Simply having its visibility set to on does not trigger it.

**Comment by WizardCM** [comment]
@cg2121 As you were the one who was able to get browser panels functional on Linux, would you have any ideas on how to move forward on this?

**Comment by mihawk90** [comment]
> The reproduction steps can be simplified just a bit:

Heh yes, I included a bunch of testing-steps just to make it obvious it works without the browser panels :D 

> * The issue will also manifest if you start up OBS with a docked browser dock.

Yes, I should have mentioned this. Originally I only tested with closing and re-opening OBS because I thought the files got corrupted on exit, it was only later I realised just closing the dock is enough.

**Comment by mihawk90** [comment]
Temporary workaround:

457adcedd319ca2317d7cd5300694d486e88af90

**Comment by GeorgesStavracas** [comment]
I've debugged this issue a bit and here are some of my findings:

 * Drag motion operates correctly; drag events are passed to scene list just as expected
 * The presence of a docked browser dock is the trigger for this behavior, but the missing drop events are _not_ routed to the dock's child window. 
 * The drop event simply vanished; it's not passed to the scene list, not even to the main window - it simply voids. My initial instinct was to look for something inside QWindowSystemInterface that could be early-returning or skipping this drop event, however [the event route is pretty direct and unchecked](https://code.woboq.org/qt5/qtbase/src/gui/kernel/qguiapplication.cpp.html#_ZN22QGuiApplicationPrivate11processDropEP7QWindowPK9QMimeDataRK6QPoint6QFlagsIN2Qt10DropActionEES8_INS9_11MouseButtonEES8_INS9_16KeyboardModifierEE).
 * I temporarily re-enabled browser docks on Wayland, and this issue is not reproducible with Qt's Wayland backend. This doesn't actually say much; it could very well be that CEF is doing something exquisite with the X11 server (like a filtered pointer grab) that won't work on Xwayland or Wayland.

These points hint that whatever is causing this issue is either in CEF, or in the Qt's xcb platform.

**Comment by RytoEX** [comment]
>     * The presence of a docked browser dock is the trigger for this behavior

I've tried to reproduce this before with just docking a browser dock, but it only seemed to occur if the browser dock _spawned_ in a docked state, not simply by docking an existing browser dock widget.  You can see this detailed explicitly in @mihawk90 's [comment above](https://github.com/obsproject/obs-studio/issues/4488#issuecomment-846266434) as well.

**Comment by tt2468** [comment]
If it helps, I've determined this to be an issue with browser dock support on linux *to begin with*. Commit 84771b143c69ccaeb2784cb63fa70e80331390a2 was what added browser docks to linux and it also has the crashing issue.

**Comment by ginsterbusch** [comment]
I have just encountered this issue, too (Kubuntu 20.04.2 LTS).
Freshly installed OBS 26, prepared scenes, sources etc, but has not seen any actual action yet (ie. not connected with Twitch etc. before).
Decided to upgrade, as the OBS upgrade on my Windows 10 Pro system didnt show any issues. 
Guess that was precipitous.

> * The presence of a docked browser dock is the trigger for this behavior
None present.

**Comment by RytoEX** [comment]
> 
> 
> I have just encountered this issue, too (Kubuntu 20.04.2 LTS).
> Freshly installed OBS 26, prepared scenes, sources etc, but has not seen any actual action yet (ie. not connected with Twitch etc. before).
> Decided to upgrade, as the OBS upgrade on my Windows 10 Pro system didnt show any issues.
> Guess that was precipitous.
> 
> > * The presence of a docked browser dock is the trigger for this behavior
> 
> None present.
> 
> I'm going to downgrade for now.

@ginsterbusch 
Why OBS 26 and not OBS 27?

Please provide a log file and _exact_ reproduction steps.  As far as we know so far, this issue is specifically triggered by code introduced for OBS 27, and it only affects Linux systems, not Windows or macOS.  Seeing this in OBS 26 would be news to us.

**Comment by kkartaltepe** [comment]
> > I have just encountered this issue, too (Kubuntu 20.04.2 LTS).
> > Freshly installed OBS 26, prepared scenes, sources etc, but has not seen any actual action yet (ie. not connected with Twitch etc. before).
> > Decided to upgrade, as the OBS upgrade on my Windows 10 Pro system didnt show any issues.
> > Guess that was precipitous.
> > > * The presence of a docked browser dock is the trigger for this behavior
> > 
> > 
> > None present.
> > I'm going to downgrade for now.
> 
> @ginsterbusch
> Why OBS 26 and not OBS 27?

Probably the User typo'ed. For anyone on release version 27, it is intentional that drag and drop does not work. This bug is to diagnose the underlying issue and why we had to disable it in 457adcedd319ca2317d7cd5300694d486e88af90 just before release.

**Comment by ginsterbusch** [comment]
> > > I have just encountered this issue, too (Kubuntu 20.04.2 LTS).
> > > Freshly installed OBS 26, prepared scenes, sources etc, but has not seen any actual action yet (ie. not connected with Twitch etc. before).
> > > Decided to upgrade, as the OBS upgrade on my Windows 10 Pro system didnt show any issues.
> > > Guess that was precipitous.
> > > > * The presence of a docked browser dock is the trigger for this behavior
> > > 
> > > 
> > > None present.
> > > I'm going to downgrade for now.
> > 
> > 
> > @ginsterbusch
> > Why OBS 26 and not OBS 27?
> 
> Probably the User typo'ed.

Re: Typo - no, there is a missing part of the paragraph. I've installed 26 from the PPA, did not do actual streaming yet, and deemed the upgrade to 27 safe, because it didnt seem to do anything seriously bad on Windows.

> For anyone on release version 27, it is intentional that drag and drop does not work. This bug is to diagnose the underlying issue and why we had to disable it in [457adce](https://github.com/obsproject/obs-studio/commit/457adcedd319ca2317d7cd5300694d486e88af90) just before release.

then that should be highlighted STRONGLY in the changelog, as this essentially makes OBS unusable for me (and probably everybody else). I wouldnt consider this a "stable" release at all (alpha at best).

Also, it cost me 4 hours of my life, that could have been better used - eg. getting my work done.

cu, w0lf.

**Comment by WizardCM** [comment]
@ginsterbusch the patch we included in 27.0 specifically disables drag and drop to avoid the bug that results in scenes being deleted. You can still reorder scenes & sources using the arrow buttons in each box.

**Comment by mihawk90** [comment]
> then that should be highlighted STRONGLY in the changelog

Well then you're in luck because it's [in the PPA changelog](https://launchpad.net/~obsproject/+archive/ubuntu/obs-studio/+packages). But I'm sure you checked that before, right?

> this essentially makes OBS unusable for me

Obviously this depends on your workflow but I'm just going to say that you can still re-arrange both scenes and sources, via the respective buttons on the lists.

> alpha at best

That very much depends on your definition of "alpha".

> wont do alpha-testing for anybody for free

Is there an impression that testers are getting payed? Where do I get my check? /s

**Comment by xurei** [comment]
The biggest problem for me is that I can't drag-and-drop sources in a group, I can only create a new group with the sources I want in.
I switched to Scene Sources as a work-around but in some cases the Group feature would be a better solution.

**Comment by GeorgesStavracas** [comment]
Spent some time debugging this, I think the fundamental issue here is that Qt5 fails to find the X Window that would receive the drop event. As a consequence, it never emits the XdndFinished event to the drag source (i.e. the lists). That ends up eating the drag data, which gets deleted, and this propagates back to OBS Studio as deleting the dragged item.

When reproducing the pathological case, the `w` variable in [`QXcbDrag::drop()` method](https://code.qt.io/cgit/qt/qtbase.git/tree/src/plugins/platforms/xcb/qxcbdrag.cpp?h=v5.15.2#n518) is null / zero. It takes the `else` clause [here](https://code.qt.io/cgit/qt/qtbase.git/tree/src/plugins/platforms/xcb/qxcbdrag.cpp?h=v5.15.2#n566), skipping the `QXcbDrag::handleDrop()` call that emits `XdndFinished`. When it's working properly, `QXcbDrag::handleDrop()` is called.

This can be verified by running OBS Studio with `QT_LOGGING_RULES=*.debug=true`. When working correctly, one can see this:

```
qt.qpa.xdnd: sending drop to target: 65011718
qt.qpa.xdnd: target: 65011718 received XdndDrop
qt.qpa.xdnd: sending XdndFinished to source: 65011716
qt.gui.dnd: mouse released over QWidgetWindow(0x564e8d6ac8d0, name="OBSBasicWindow") after drag from QWindow(0x0) globalPos QPoint(2352,1096)
```

and when reproducing this issue, the following is logged:

```
qt.qpa.xdnd: sending drop to target: 65011718
qt.gui.dnd: mouse released over QWidgetWindow(0x559d7f324df0, name="OBSBasicWindow") after drag from QWindow(0x0) globalPos QPoint(2336,1123)
```

Now, the question is: what causes the `w` variable to be null? I can see two ways this could happen: (1) either the `connection()->platformWindowFromId(current_proxy_target)` line returns null, or (2) the if clause just after it sets it to zero. This condition is:

```
    if (w && w->window()->type() == Qt::Desktop)
        w = nullptr;
```

Since none of the X Windows in OBS Studio are certainly not desktop windows, then we're left only with option (1). I didn't figure if (1) is returning null because `current_proxy_target` is `XCB_NONE` (which is zero), or if it's a valid value but no matching platform window is found. This is where I'm currently stuck, if someone knows X11, please please please help.

I hate X.

**Comment by DiPiVR** [comment]
Sorry to necro this issue, but I have encountered this bug on OBS 32.1.2, on Linux Mint 22.3 (Ubuntu 24.04 base), installed using this [PPA](https://launchpad.net/~obsproject/+archive/ubuntu/obs-studio).

Repro steps:
1. launch OBS with twitch chat dock enabled. Dragging doesn't work.
2. Hide chat dock. Relaunch, dragging works.
3. Show chat dock -> dragging stops working.

OBS also crashed without leaving anything in log when I moved few scenes around and watched them disappear, but that _might_ be because I did in [Scene Tree](https://github.com/TheThirdRail/scene-tree-view/) dock...
